### PR TITLE
chore(jellyfish): add negative interest rate tests

### PR DESCRIPTION
**Pull request by DieHard073055**
#### What this PR does / why we need it:
Tests the negative interest rate update that was released.

#### Which issue(s) does this PR fixes?:

Fixes part of #1749

#### Additional comments?:
I will be pushing smaller changes at a time, while i get reviews on the work.

TODO: 
- [x]  should take a loan with negative interest rate and accrue DUSD
- [x]  should fully pay back a vault with negative interest rate
- [x]  should fully pay back a vault with multiple loans including negative interest rate loan
- [x] should pay back a vault partially with negative interest rate
- [x] should pay back a vault partially with multiple loans including negative interest rate loan

### CODE REVIEW

File: packages/testcontainers/src/containers/RegTestContainer/index.ts [hunk 1]
```diff
@@ -37,7 +37,9 @@ export class RegTestContainer extends DeFiDContainer {
       '-fortcanninghillheight=10',
       '-fortcanningroadheight=11',
       '-fortcanningcrunchheight=12',
-      '-fortcanningspringheight=13'
+      '-fortcanningspringheight=13',
+      'fortcanninggreatworldheight=14',
```
Comment (DieHard073055) [inline packages/testcontainers/src/containers/RegTestContainer/index.ts]: '-fortcanninggreatworldheight=14',
'-negativeinterest=1'

Seems like these 2 flags are causing the container to stop.

**Comment by DieHard073055** [comment]
Now when I take a loan with
```javascript
    const dusdLoanAmount = 5000
    const txid = await testing.rpc.loan.takeLoan({
      vaultId: bobVaultId,
      amounts: `${dusdLoanAmount}@DUSD`
    })
```
And then check the interest rate with
```javascript
    const interests = await testing.rpc.loan.getInterest('scheme')
```
The interest is 0
```javascript
    console.log(interests[0].totalInterest.toFixed(8)) // 0.00000000
    console.log(interests[0].interestPerBlock.toFixed(8)) // 0.00000000
    console.log(interests[0].realizedInterestPerBlock.toFixed(8)) // 0.00000000
```

**Comment by DieHard073055** [comment]
But if I comment out the setGov function, I get some value for the interest rate
```javascript
    await testing.rpc.masternode.setGov({
      ATTRIBUTES: {
        'v0/token/1/loan_minting_interest': '-5'
      }
    })
    await testing.generate(1)
```
the result is something like
```javascript
    console.log(interests[0].totalInterest.toFixed(8)) // 0.00475647
    console.log(interests[0].interestPerBlock.toFixed(8)) // 0.00475647
    console.log(interests[0].realizedInterestPerBlock.toFixed(8)) // 0.00475647
```

File: packages/testcontainers/src/containers/RegTestContainer/index.ts (hunk 1 above)
Comment (Jouzo) [reply packages/testcontainers/src/containers/RegTestContainer/index.ts]: `-negativeinterest=1` is used on ain to track negative interest, not activate it. See https://github.com/DeFiCh/ain/blob/28f2b9ea34487a087a6de82d2b0004f559a69d77/src/masternodes/govvariables/attributes.cpp#L415

**Comment by DieHard073055** [comment]
```
 FAIL  packages/jellyfish-api-core/__tests__/category/loan/negativeInterest.test.ts (38.484 s)                         
  takeLoan with negative interest success                                                                              
    ✓ should takeLoan with negative interest and accrue DUSD (18044 ms)                                                
    ✕ should fully pay back a vault with negative interest rate (17752 ms)                                             
                                                                                                                       
  ● takeLoan with negative interest success › should fully pay back a vault with negative interest rate                
                                                                                                                       
    RpcApiError: 'Test PaybackLoanTx execution failed:                                                                 
    amount 0.00000000 is less than 4995.24353121', code: -32600, method: paybackloan 
```
I wonder what I am doing wrong here.
```javascript
 // payback loan
    const loanPaybackResult = await testing.rpc.loan.paybackLoan({
      vaultId: bobVaultId,
      amounts: afterLoanAmount, // console.log shows the value is 4995.24353121@DUSD
      from: aliceAddr
    })
```

**Comment by DieHard073055** [comment]
I have implemented all the tests.
I believe I need to calculate all the numbers in the test, rather then defining a variable with a magic number and then asserting against that ?

File: packages/jellyfish-api-core/__tests__/category/loan/negativeInterest.test.ts [hunk 2]
```diff
@@ -0,0 +1,494 @@
...
+
+let aliceAddr: string
+let bobVaultId: string
+let bobVaultAddr: string
+let bobVault: VaultActive
+let oracleId: string
+let timestamp: number
+const blocksPerDay = (60 * 60 * 24) / (10 * 60) // 144 in regtest
+// High interest rate so the change will be significant
+const interestRate = 5000
```
Comment (canonbrother) [inline packages/jellyfish-api-core/__tests__/category/loan/negativeInterest.test.ts:16]: hmm.. `5000` is quite abnormal.. preferably provide a more reality interest rate.. maybe `3.3`??

Comment (DieHard073055) [reply packages/jellyfish-api-core/__tests__/category/loan/negativeInterest.test.ts:16]: Yea, i wanted to see a significant change with just one block added to the chain.

**Comment by canonbrother** [comment]
heyy @DieHard073055 
that's some failed test.. can you check if that is regression issue??

**Comment by fuxingloh** [comment]
I’ve a PR that fixes the regression but I haven’t see if it’s all fixed thereRegards,Fuxing LohOn 15 Sep 2022, at 12:29 PM, canonbrother ***@***.***> wrote:﻿
heyy @DieHard073055
that's some failed test.. can you check if that is regression issue??

**Comment by DieHard073055** [comment]
One of the errors is happening in [setLoanToken.test.ts](https://github.com/JellyfishSDK/jellyfish/blob/DieHard073055/add_negative_interest_rate_tests/packages/jellyfish-api-core/__tests__/category/loan/setLoanToken.test.ts#L324-L329)

Because of the following condition `interest rate cannot be less than 0!`
We now allow this.

Should I remove this test? @fuxingloh

**Comment by canonbrother** [comment]
> I’ve a PR that fixes the regression but I haven’t see if it’s all fixed there

Noted that

@DieHard073055 
just want make sure whether its regression or flaky <-- you can try run it local to verify... 
overall the pr look good to me

File: packages/jellyfish-api-core/__tests__/category/loan/negativeInterest.test.ts (hunk 2 above)
Comment (canonbrother) [reply packages/jellyfish-api-core/__tests__/category/loan/negativeInterest.test.ts:16]: noted that.. hmm.. the reason why i recommend to provide tricky figure like `3.3` or `5.6433` is to check the precision working
the `significant change ` can be tested by `c.generate(36)`

**Comment by fuxingloh** [comment]
> One of the errors is happening in [setLoanToken.test.ts](https://github.com/JellyfishSDK/jellyfish/blob/DieHard073055/add_negative_interest_rate_tests/packages/jellyfish-api-core/__tests__/category/loan/setLoanToken.test.ts#L324-L329)
> 
> Because of the following condition `interest rate cannot be less than 0!` We now allow this.
> 
> Should I remove this test? @fuxingloh

If they are no longer valid `it.skip` them or fix them if possible.

File: packages/testcontainers/src/containers/RegTestContainer/index.ts [hunk 3]
```diff
@@ -37,7 +37,8 @@ export class RegTestContainer extends DeFiDContainer {
       '-fortcanninghillheight=10',
       '-fortcanningroadheight=11',
       '-fortcanningcrunchheight=12',
-      '-fortcanningspringheight=13'
+      '-fortcanningspringheight=13',
+      '-greatworldheight=14'
```
Comment (fuxingloh) [inline packages/testcontainers/src/containers/RegTestContainer/index.ts:41]: This flag was introduced for `v2.10.0`?

```suggestion
      '-fortcanninggreatworldheight=14'
```

Also, can we set the full flag using the full name?

**Comment by fuxingloh** [comment]
@DieHard073055 can you also update the 2 `docker-compose.yml` with the new flag.

File: packages/testcontainers/src/containers/RegTestContainer/index.ts (hunk 3 above)
Comment (DieHard073055) [reply packages/testcontainers/src/containers/RegTestContainer/index.ts:41]: This breaks the test. The container just stops. I am not sure why.

Comment (DieHard073055) [reply packages/testcontainers/src/containers/RegTestContainer/index.ts:41]: @canonbrother might know why.

Comment (canonbrother) [reply packages/testcontainers/src/containers/RegTestContainer/index.ts:41]: yea.. in `v2.10.0` called it `greatworldheight`.. after that get renamed to `fortcanninggreatworldheight`
https://github.com/DeFiCh/ain/pull/1435/files

Comment (fuxingloh) [reply packages/testcontainers/src/containers/RegTestContainer/index.ts:41]: Oh! Thanks!

**Review by canonbrother — APPROVED** [review]

**Review by fuxingloh — APPROVED** [review]

**Pull request merged**